File: src/main/java/hdzi/editstarters/ui/EditStartersDialog.java
package hdzi.editstarters.ui;

import com.intellij.openapi.actionSystem.CommonDataKeys;
import com.intellij.openapi.command.WriteCommandAction;
import com.intellij.ui.CollectionComboBoxModel;
import com.intellij.ui.CollectionListModel;
import hdzi.editstarters.buildsystem.BuildSystem;
import hdzi.editstarters.dependency.Module;
import hdzi.editstarters.dependency.*;
import hdzi.editstarters.initializr.InitializrStatus;
import hdzi.editstarters.version.Version;
import org.apache.commons.lang.WordUtils;

import javax.swing.*;
import java.awt.event.*;
import java.text.SimpleDateFormat;
import java.util.*;
import java.util.stream.Collectors;

import static javax.swing.WindowConstants.DO_NOTHING_ON_CLOSE;

public class EditStartersDialog {
    private JPanel root;
    private JButton buttonOK;
    private JButton buttonCancel;
    private JComboBox<Version> versionComboBox;
    private JList<String> moduleList;
    private JList<StarterInfo> starterList;
    private JList<StarterInfo> selectList;
    private JTextField searchField;
    private JCheckBox cachedBox;
    private JCheckBox ohubBox;
    private final JFrame frame;
    private final Set<StarterInfo> addStarters = new HashSet<>(64);
    private final Set<StarterInfo> removeStarters = new HashSet<>(64);
    private final WeakHashMap<StarterInfo, String> toolTipTextCache = new WeakHashMap<>(); // 加个缓存
    private final WeakHashMap<StarterInfo, String> searchCache = new WeakHashMap<>(); // 搜索缓存

    public EditStartersDialog(BuildSystem buildSystem, SpringBoot springBoot, InitializrStatus status) {
        this.frame = new JFrame("Edit Starters");
        this.frame.setContentPane(this.root);

        // 是否使用缓存
        if (status.isEnableCache()) {
            this.cachedBox.setSelected(true);
            this.cachedBox.addMouseMotionListener(new MouseAdapter() {
                @Override
                public void mouseMoved(MouseEvent e) {
                    cachedBox.setToolTipText(new SimpleDateFormat("yyyy-MM-dd HH:mm:ss")
                            .format(new Date(status.getCacheUpdateTime())));
                }
            });
        }

        // 是否启用Ohub
        if (status.isEnableOHub()) {
            this.ohubBox.setSelected(true);
            this.ohubBox.addMouseMotionListener(new MouseAdapter() {
                @Override
                public void mouseMoved(MouseEvent e) {
                    ohubBox.setToolTipText(status.getOHubName());
                }
            });
        }

        // boot版本选框
        this.versionComboBox.setModel(new CollectionComboBoxModel<>(
//                springBoot.getVersion().getValues().stream().map(InitializrVersion.Value::getId).collect(Collectors.toList()),
                Collections.singletonList(springBoot.getBootVersion()),
                springBoot.getBootVersion()
        ));
        this.versionComboBox.setEnabled(false);

        // 取消按钮
        this.buttonCancel.addActionListener(e -> onCancel());
        // 点击 X 时调用 onCancel()
        this.frame.setDefaultCloseOperation(DO_NOTHING_ON_CLOSE);
        this.frame.addWindowListener(new WindowAdapter() {
            public void windowClosing(WindowEvent e) {
                onCancel();
            }
        });
        // 遇到 ESCAPE 时调用 onCancel()
        this.root.registerKeyboardAction(e -> onCancel(),
                KeyStroke.getKeyStroke(KeyEvent.VK_ESCAPE, 0), JComponent.WHEN_ANCESTOR_OF_FOCUSED_COMPONENT);


        // ok按钮
        this.frame.getRootPane().setDefaultButton(this.buttonOK);
        this.buttonOK.addActionListener(e -> {
            WriteCommandAction.runWriteCommandAction(buildSystem.getContext().getData(CommonDataKeys.PROJECT), () -> {
                buildSystem.addStarters(this.addStarters);
                buildSystem.removeStarters(this.removeStarters);
            });
            this.frame.dispose();
        });

        Map<String, List<StarterInfo>> modules = springBoot.getModules().stream()
                .collect(Collectors.toMap(Module::getName, Module::getValues, (o, n) -> o, LinkedHashMap::new));
        // Module列表
        this.moduleList.setModel(new CollectionListModel<>(modules.keySet()));
        this.moduleList.addMouseListener(new MouseAdapter() {
            @Override
            public void mouseClicked(MouseEvent e) {
                searchField.setText("");
                String name = moduleList.getSelectedValue();
                starterList.setModel(new CollectionListModel<>(modules.getOrDefault(name, Collections.emptyList())));
            }
        });

        // 显示详细信息
        MouseAdapter showDescAdapter = new MouseAdapter() {
            @Override
            public void mouseMoved(MouseEvent e) {
                @SuppressWarnings("unchecked")
                JList<StarterInfo> list = (JList<StarterInfo>) e.getSource();
                int index = list.locationToIndex(e.getPoint());
                if (index > -1) {
                    StarterInfo starter = list.getModel().getElementAt(index);
                    list.setToolTipText(getStarterInfoToolTipText(starter));
                }
            }
        };

        List<Dependency> existDependencies = buildSystem.getDependencies();
        // Starter列表
        this.starterList.addMouseListener(new MouseAdapter() {
            @Override
            public void mouseClicked(MouseEvent e) {
                StarterInfo starterInfo = starterList.getSelectedValue();
                if (Points.contains(existDependencies, starterInfo)) { // 对于已存在的starter，添加就是从删除列表里删除
                    removeStarters.remove(starterInfo);
                } else { // 对于不存在的starter，添加直接加入添加列表
                    addStarters.add(starterInfo);
                }
                // 去重显示
                CollectionListModel<StarterInfo> listModel = (CollectionListModel<StarterInfo>) selectList.getModel();
                if (!listModel.contains(starterInfo)) {
                    listModel.add(starterInfo);
                }
            }
        });
        this.starterList.addMouseMotionListener(showDescAdapter);

        // selected列表
        List<StarterInfo> existStarters = modules.values().stream()
                .flatMap(List::stream)
                .filter(info -> Points.contains(existDependencies, info))
                .collect(Collectors.toList());
        this.selectList.setModel(new CollectionListModel<>(existStarters));
        this.selectList.addMouseListener(new MouseAdapter() {
            @Override
            public void mouseClicked(MouseEvent e) {
                StarterInfo starterInfo = selectList.getSelectedValue();
                if (Points.contains(existDependencies, starterInfo)) { // 对于已存在的starter，删除就是加入删除列表
                    removeStarters.add(starterInfo);
                } else { // 对于不存在的starter，删除是从添加列表里删除
                    addStarters.remove(starterInfo);
                }
                // 显示
                ((CollectionListModel<StarterInfo>) selectList.getModel()).remove(starterInfo);
            }
        });
        this.selectList.addMouseMotionListener(showDescAdapter);

        // 搜索框
        this.searchField.addKeyListener(new KeyAdapter() {
            @Override
            public void keyReleased(KeyEvent e) {
                moduleList.clearSelection();
                String searchKey = searchField.getText().toLowerCase();
                List<StarterInfo> result = modules.values().stream().flatMap(starters -> starters.stream().filter(starter ->
                                searchCache.computeIfAbsent(starter, key -> (key.getGroupId() + ":" + key.getArtifactId() + "\t" + key.getName()).toLowerCase())
                                        .contains(searchKey)))
                        .collect(Collectors.toList());
                starterList.setModel(new CollectionComboBoxModel<>(result));
            }
        });
    }

    private void onCancel() {
        this.frame.dispose();
    }

    public void show() {
        this.frame.pack();
        this.frame.setLocationRelativeTo(null); // 中间显示
        this.frame.setVisible(true);
    }

    private String getStarterInfoToolTipText(StarterInfo starter) {
        return toolTipTextCache.computeIfAbsent(starter, info -> {
            StringBuilder buffer = new StringBuilder();
            buffer.append("Name: ").append(info.getName()).append("<br/>")
                    .append("GroupId: ").append(info.getGroupId()).append("<br/>")
                    .append("ArtifactId: ").append(info.getArtifactId()).append("<br/>")
                    .append("Scope: ").append(info.getScope()).append("<br/>");
            if (info.getVersion() != null) {
                buffer.append("Version: ").append(info.getVersion()).append("<br/>");
            }
            if (info.getVersionRange() != null) {
                buffer.append("Version Range: ").append(info.getVersionRange()).append("<br/>");
            }
            buffer.append("<br/>").append(WordUtils.wrap(info.getDescription(), 50, "<br/>", false));

            return buffer.toString();
        });
    }
}

File: src/main/kotlin/hdzi/editstarters/ui/EditStartersDialog.kt
package hdzi.editstarters.ui

import com.intellij.openapi.actionSystem.CommonDataKeys
import com.intellij.openapi.command.WriteCommandAction
import com.intellij.openapi.project.Project
import com.intellij.ui.CollectionComboBoxModel
import com.intellij.ui.CollectionListModel
import hdzi.editstarters.buildsystem.BuildSystem
import hdzi.editstarters.springboot.initializr.SpringInitializr
import hdzi.editstarters.springboot.initializr.StarterInfo
import org.apache.commons.lang.WordUtils
import java.awt.event.KeyAdapter
import java.awt.event.KeyEvent
import java.awt.event.MouseAdapter
import java.awt.event.MouseEvent
import java.util.*
import javax.swing.*

class EditStartersDialog(buildSystem: BuildSystem, initializr: SpringInitializr) {
    private lateinit var root: JPanel
    private lateinit var buttonOK: JButton
    private lateinit var buttonCancel: JButton
    private lateinit var versionComboBox: JComboBox<String>
    private lateinit var moduleList: JList<String>
    private lateinit var starterList: JList<StarterInfo>
    private lateinit var selectList: JList<StarterInfo>
    private lateinit var searchField: JTextField
    private val frame: JFrame
    private val title = "Edit Starters"
    private val addStarters = HashSet<StarterInfo>(64)
    private val removeStarters = HashSet<StarterInfo>(64)
    private val toolTipTextCache = WeakHashMap<StarterInfo, String>() // 加个缓存

    init {
        this.frame = JFrame(this.title)
        this.frame.contentPane = this.root

        // boot版本选框
        this.versionComboBox.model = CollectionComboBoxModel(
            initializr.version.values.map { it.id },
            initializr.currentVersionID
        )
        this.versionComboBox.isEnabled = false

        // 取消按钮
        this.buttonCancel.addActionListener { this.frame.dispose() }

        // ok按钮
        this.buttonOK.addActionListener {
            WriteCommandAction.runWriteCommandAction(buildSystem.context.getData<Project>(CommonDataKeys.PROJECT)) {
                buildSystem.addStarters(this.addStarters)
                buildSystem.removeStarters(this.removeStarters)
            }
            this.frame.dispose()
        }

        val modulesMap = initializr.modulesMap

        // Module列表
        this.moduleList.model = CollectionListModel(modulesMap.keys)
        this.moduleList.addMouseListener(object : MouseAdapter() {
            override fun mouseClicked(e: MouseEvent) {
                searchField.text = ""
                val name = moduleList.selectedValue
                starterList.model = CollectionListModel(modulesMap[name] ?: Collections.emptyList())
            }
        })

        // 显示详细信息
        val showDescAdapter = object : MouseAdapter() {
            override fun mouseMoved(e: MouseEvent) {
                @Suppress("unchecked_cast")
                val list = e.source as JList<StarterInfo>
                val index = list.locationToIndex(e.point)
                if (index > -1) {
                    val starter = list.model.getElementAt(index)
                    list.toolTipText = starter.getStarterInfoToolTipText()
                }
            }
        }

        // Starter列表
        this.starterList.addMouseListener(object : MouseAdapter() {
            override fun mouseClicked(e: MouseEvent) {
                if (e.clickCount == 2) { // 按两下选择
                    val starterInfo = starterList.selectedValue
                    if (starterInfo.exist) { // 对于已存在的starter，添加就是从删除列表里删除
                        removeStarters.remove(starterInfo)
                    } else { // 对于不存在的starter，添加直接加入添加列表
                        addStarters.add(starterInfo)
                    }
                    // 去重显示
                    val listModel = selectList.model as CollectionListModel<StarterInfo>
                    if (!listModel.contains(starterInfo)) {
                        listModel.add(starterInfo)
                    }
                }
            }
        })
        this.starterList.addMouseMotionListener(showDescAdapter)

        // selected列表
        this.selectList.model = CollectionListModel(initializr.existStarters)
        this.selectList.addMouseListener(object : MouseAdapter() {
            override fun mouseClicked(e: MouseEvent) {
                if (e.clickCount == 2) { // 按两下删除
                    val starterInfo = selectList.selectedValue
                    if (starterInfo.exist) { // 对于已存在的starter，删除就是加入删除列表
                        removeStarters.add(starterInfo)
                    } else { // 对于不存在的starter，删除是从添加列表里删除
                        addStarters.remove(starterInfo)
                    }
                    // 显示
                    (selectList.model as CollectionListModel<StarterInfo>).remove(starterInfo)
                }
            }
        })
        this.selectList.addMouseMotionListener(showDescAdapter)

        // 搜索框
        this.searchField.addKeyListener(object : KeyAdapter() {
            override fun keyReleased(e: KeyEvent) {
                moduleList.clearSelection()
                val searchKey = searchField.text.toLowerCase()
                val result = modulesMap.values.flatMap { starters ->
                    starters.filter { starter ->
                        "${starter.groupId}:${starter.artifactId}\t${starter.name}"
                            .toLowerCase().contains(searchKey)
                    }
                }
                starterList.model = CollectionComboBoxModel(result)
            }
        })
    }

    fun show() {
        this.frame.pack()
        this.frame.setLocationRelativeTo(null) // 中间显示
        this.frame.isVisible = true
    }

    private fun StarterInfo.getStarterInfoToolTipText(): String =
        toolTipTextCache.computeIfAbsent(this) { info ->
            val buffer = StringBuilder()
            buffer.append("GroupId: ").append(info.groupId).append("<br/>")
                .append("ArtifactId: ").append(info.artifactId).append("<br/>")
                .append("Scope: ").append(info.scope).append("<br/>")
            if (info.version != null) {
                buffer.append("Version: ").append(info.version).append("<br/>")
            }
            if (info.versionRange != null) {
                buffer.append("Version Range: ").append(info.versionRange).append("<br/>")
            }
            buffer.append("<br/>").append(WordUtils.wrap(info.description, 50, "<br/>", false))

            buffer.toString()
        }
}
